### ISSUE TITLE
Field order mismatch after serialization and deserialization

### ISSUE THREAD

**Issue by arvgord** [FasterXML/jackson-databind#4751]
### Search before asking

- [X] I searched in the [issues](https://github.com/FasterXML/jackson-databind/issues) and found nothing similar.

### Describe the bug

I am encountering a problem where the field order is not preserved after serializing and deserializing objects in Jackson.

I need access to only one specific field (transactionId), while I want to pass the remaining fields as they are, preserving the order in which they appear. I expected the transactionId field order might change, but I did not expect the order of the other fields `(a, b, c)` to change, especially in the second and third tests (SecondObject and ThirdObject). The data inside the `c` array is preserved in the expected order, but the surrounding fields are re-ordered.

### Version Information

2.17.2

## Tested Objects:
```kotlin
@JsonAutoDetect(
    fieldVisibility = JsonAutoDetect.Visibility.ANY,
    getterVisibility = JsonAutoDetect.Visibility.NONE
)
class FirstObject {
    @field:JsonAnySetter
    @field:JsonAnyGetter
    val data: Map<String, Any?> = LinkedHashMap()

    val transactionId: String
        get() = data["transactionId"] as String
}

@JsonAutoDetect(
    fieldVisibility = JsonAutoDetect.Visibility.ANY,
    getterVisibility = JsonAutoDetect.Visibility.NONE
)
class SecondObject(
    val transactionId: String
) {
    @field:JsonAnySetter
    @field:JsonAnyGetter
    val data: Map<String, Any?> = LinkedHashMap()
}

@JsonAutoDetect(
    fieldVisibility = JsonAutoDetect.Visibility.ANY,
    getterVisibility = JsonAutoDetect.Visibility.NONE
)
class ThirdObject(
    val transactionId: String,
    @field:JsonAnySetter
    @field:JsonAnyGetter
    val data: Map<String, Any?> = LinkedHashMap()
)
``` 
## Input JSON for tests:
```json
{
  "b": 2,
  "a": 1,
  "transactionId": "test",
  "c": [
    {
      "id": "3",
      "value": "c"
    },
    {
      "id": "1",
      "value": "a"
    },
    {
      "id": "2",
      "value": "b"
    }
  ]
}
```
## Test Case:
Here’s a test case to reproduce the issue. The following Kotlin test serializes and deserializes three different classes (FirstObject, SecondObject, and ThirdObject), comparing the serialized JSON with the original input JSON.
[JacksonSortingTest.kt](https://github.com/arvgord/jackson-databind-sorting-issue/blob/main/src/test/kotlin/JacksonSortingTest.kt).
### First Test Result (Passes):
The first test for FirstObject passes as expected, with the field order being preserved.
### Failing Test Results:
Expected:
```json
{
  "b": 2,
  "a": 1,
  "transactionId": "test",
  "c": [
    {
      "id": "3",
      "value": "c"
    },
    {
      "id": "1",
      "value": "a"
    },
    {
      "id": "2",
      "value": "b"
    }
  ]
}
```
For SecondObject was:
```json
{
  "transactionId": "test",
  "a": 1,
  "b": 2,
  "c": [
    {
      "id": "3",
      "value": "c"
    },
    {
      "id": "1",
      "value": "a"
    },
    {
      "id": "2",
      "value": "b"
    }
  ]
}
```
For ThirdObject was:
```json
{
  "transactionId": "test",
  "c": [
    {
      "id": "3",
      "value": "c"
    },
    {
      "id": "1",
      "value": "a"
    },
    {
      "id": "2",
      "value": "b"
    }
  ],
  "a": 1,
  "b": 2
}
```
Repository for Reproduction:
You can find a repository with the full reproduction of the issue at [jackson-databind-sorting-issue](https://github.com/arvgord/jackson-databind-sorting-issue).

### Expected behavior

For SecondObject and ThirdObject, after serializing the object back to JSON, I expect the `transactionId` field to appear at the top, followed by the fields `b`, `a`, and `c`, in the exact order they were present when the JSON was deserialized into SecondObject and ThirdObject.

**Comment by arvgord** [comment]
I tried to update from 2.17.2 to 2.18.0 but found that serialization was broken. Issue #4752

**Comment by cowtowncoder** [comment]
Same as with other issue:

* If test can be made Java-only, can remain here;
* if not, needs to be moved to `jackson-module-kotlin`

**Comment by cowtowncoder** [comment]
Could you please include test input JSON inline in description, and not just in included (Kotlin) test? It is hard to follow description wrt "a", "b" and "c" when Objects do not have such fields -- I assume they are for "any properties" from JSON.

**Comment by arvgord** [comment]
> Could you please include test input JSON inline in description, and not just in included (Kotlin) test? It is hard to follow description wrt "a", "b" and "c" when Objects do not have such fields -- I assume they are for "any properties" from JSON.

I have now included the test input JSON directly in the issue description for better clarity.

To clarify, I expect that all data from the JSON, which is not explicitly defined as `transactionId` field, will be set in the data map using `@JsonAnySetter`.

**Comment by arvgord** [comment]
I created tests for Java: https://github.com/arvgord/jackson-databind-sorting-issue-java. For Jackson version 2.17.2, everything works fine. The field order is preserved as I expected:
```json
{
    "transactionId": "test",
    "b": 2,
    "a": 1,
    "c": [
        {
            "id": "3",
            "value": "c"
        },
        {
            "id": "1",
            "value": "a"
        },
        {
            "id": "2",
            "value": "b"
        }
    ]
}
```
It seems that the issue is related to [jackson-module-kotlin #842](https://github.com/FasterXML/jackson-module-kotlin/issues/842). Therefore, this issue can be closed.

**Comment by arvgord** [comment]
I can reproduce this issue with Java, so I reopened it. Here’s the [test](https://github.com/arvgord/jackson-databind-sorting-issue-java/blob/main/src/test/java/com/arvgord/JacksonSortingTest.java) to reproduce.

**Comment by pjfanning** [comment]
Just because the code used to work a certain way and now doesn't work that way - that doesn't mean the old way was intended or that the change is a regression or a bug.

With Serialization, the one change that I can see in the [release notes](https://github.com/FasterXML/jackson-databind/blob/2.19/release-notes/VERSION-2.x) is #4580.

I don't even know how Jackson is supposed to preserve the order. When you deserialize to Java object instance and then serialize the Java object instance - there is currently nothing stored on the Java object instance to say what the order of the original Java was.

**Comment by arvgord** [comment]
I'm encountering different field orders when serializing objects, which is confusing. It would be helpful if Jackson supported a unified field order by default, perhaps based on the order of fields in the constructor. Consistency in field order would improve predictability and make it easier to manage serialization across various use cases. Is there any chance of supporting a standardized order as the default in future versions?

**Comment by pjfanning** [comment]
Read https://github.com/FasterXML/jackson-databind/issues/4580

That is predicatable, isn't it?

**Comment by cowtowncoder** [comment]
Field order really should not be relied on by processors, and so while we do try to make an effort to keep things both predictable and compatible across versions, like @pjfanning suggested this is not always possible. And specifically, sometimes old behavior is considered worth changing.
Note, in particular, that Records do have more predictable ordering than POJOs.

I'll need to read this in more detail, above are my initial quick thoughts.

**Comment by cowtowncoder** [comment]
@arvgord Ok so wrt test you linked to (https://github.com/arvgord/jackson-databind-sorting-issue-java/blob/main/src/test/java/com/arvgord/JacksonSortingTest.java) looks like:

1. One of 3 tests ("First") works as-is
2. Two other tests fail due to ordering, but have same logical contents

Due to complexity of the case -- use of "any" getters, sets, not specifying `@JsonPropertyOrder` or `MapperFeature.SORT_PROPERTIES_ALPHABETICALLY` (or `MapperFeature.SORT_CREATOR_PROPERTIES_FIRST`) -- there are not many guarantees on ordering. This because serialization order of properties is not (and cannot be) guaranteed when it is based on arbitrary order in which Class Fields and Methods are made available by JDK introspection.

But there are some Jackson defaults that may be helpful to know. Ordering goes like this:

1. Properties listed on `@JsonPropertyOrder` come BEFORE all other properties
2. Creator properties (properties that are bound on deserialization to Creators) are sorted BEFORE other properties (unless `MapperFeature.SORT_CREATOR_PROPERTIES_FIRST` is disabled)
3. Regular properties that are not covered by (1) or (2) come next
4. "Any" properties are sorted AFTER all other properties

This is the expected ordering for 2.18, as far as I know.

With that; if there are specific remaining aspects of 3 tests that still seem wrong with 2.18.1, let's outline those. That ordering differs from 2.17 is in itself not a flaw; but if the ordering seems incorrect based on my listing above can point to a flaw.

I will tentatively close this issue: if remaining issues exist as per above, I'll re-open this.
